lib/domain.js

```diff
@@ -12,26 +9,43 @@ var getRulesForTld = require('./tld-rules.js');
  * @param {string} host
  * @return {String}
  */
-module.exports = function getDomain (allRules, validHosts, host) {
-  var domain = null, hostTld, rules, rule;
+module.exports = function getDomain(allRules, validHosts, host, isHostClean) {
   var _validHosts = validHosts || [];
+  var cleanHost = cleanHostValue(host, isHostClean);
 
-  if (isValid(_validHosts, host) === false) {
+  if (isValid(_validHosts, cleanHost) === false) {
     return null;
   }
 
-  host = cleanHostValue(host);
-  hostTld = extractTldFromHost(host);
-  rules = getRulesForTld(allRules, hostTld, new Rule({"firstLevel": hostTld, "isHost": _validHosts.indexOf(hostTld) !== -1}));
-  rule = getCandidateRule(host, rules);
+  // Check if `host` ends with '.' followed by one host specified in validHosts.
+  for (var i = 0; i < validHosts.length; i++) {
+    var vhost = validHosts[i];
+    if (cleanHost.indexOf(vhost) === (cleanHost.length - vhost.length) && (
+          cleanHost.length === vhost.length ||
+          cleanHost[cleanHost.length - vhost.length - 1] === '.')) {
+      return vhost;
+    }
+  }
 
-  if (rule === null) {
+  var suffix = getPublicSuffix(allRules, cleanHost, true);
+  if (suffix === null) {
+    // TODO - shouldn't it be null?
+    // Otherwise 'should return the known valid host' fails
+    // return cleanHost;
     return null;
   }
 
-  host.replace(new RegExp(rule.getPattern()), function (m, d) {
-    domain = d;
-  });
+  if (suffix.length === cleanHost.length) {
+    return null;
+  }
+
+  // google.fr (length 9)
+  // suffix = fr (length 2)
+  // 5 = 9 - 2 - 1 (ignore the dot) - 1 (zero-based indexing)
+  var lastDotBeforeSuffixIndex = cleanHost.lastIndexOf('.', cleanHost.length - suffix.length - 2);
```

Comment:
Don't hesitate to make it a function that returns `null` or a `CleanHostString`.